### DISCUSSION TITLE
One File Upload Solution

### DISCUSSION THREAD

**Discussion by noor-codes** [keystonejs/keystone#7457] [Feature requests]
# One File Upload Solution

Keystone offers two good solutions for uploading files. [Images](https://keystonejs.com/docs/apis/fields#image) and [Files](https://keystonejs.com/docs/apis/fields#file)

### This seems to be a good solution for most people but it doesn't cover the edge cases and often comes with drawbacks such as:

1: Files-related data goes to the files table and the Images related data can go to the images table.
2: Instead of maintaining 1 field, two different fields are maintained. 
3: What If you have 1 table where all the files are stored and still, If it's an image, it should show a thumbnail otherwise just preview the link to the file. 

I went through a lot to get the result in the video but it would be awesome if support for such a possibility lands by default.
I am also eager to know about the decision behind the separation of these two fields. 

Thank You :)

https://user-images.githubusercontent.com/28804764/162708717-01564418-34d2-446c-afe3-ed108f67ceb9.mp4

**Comment by pc-erin** [comment]
We could also use a feature like this.

We're trying to create a featured media list to associate with items in other lists, but the featured media could be images or videos.

This means we either have to make multiple fields, one for image and one of type "file" for videos, or we have to make everything a "file".

Looking through the [file upload code](https://github.com/keystonejs/keystone/tree/main/packages/core/src/lib/assets) I notice the main difference between them is the mime types added to the S3 object, and the image context does some metadata parsing.

My first thought was to just add my own `createVideoContext` function and make another storage type "video", but it would be nice to have a more general solution for all file types, and one that doesn't require making a separate field for every file type you want to support.

What if we just had one `createFileContext` function which would parse metadata and set MIME headers based on the file extension? It could have a built-in list of supported file types, their associated metadata type definitions, and metadata extraction functions. We could also have `StorageConfig` support using custom extensions by specifying MIME type, metadata format, and extraction function for each one. Might also be good to have an option to restrict the built-in types that are supported on a storage zone in case you want to make a zone that is only for text files, etc.

I think that would be the most flexible way of supporting any file types someone wants to use, but also keeping the type-safety we've come to expect from Keystone.

At that point the different field types that interact with storage would just be different frontends for interacting with backend files that all have the same shape except for their `type` and `metadata`.
